File: node/firsts.js
/**
   Downloaders
**/
var http = require("http");
//var async = require("async");
var Q = require("q")

function downloadAll(subreddits, finalCallback) {
    return Q.all(subreddits.map(function(subreddit) {
	return downloadSubReddit(subreddit).then(
	    function(mbLinks) {
		return Q.all(
		    mbLinks.getDefault([]).map(downloadFirst)
		);
	    }
	)
    })).then(function(results) { 
	return concatMap(function(x) { return x; }, results);
    });
}
exports.downloadAll = downloadAll;

function concatMap(f, arr) {
    return Array.prototype.concat.apply([], arr.map(f))
}

function downloadBody(url) {
    var deferred = Q.defer();
    http.get(url, function(res) {
	var body = ""

	res.on("data", function(chunk) {
	    body += chunk;
	});

	res.on("end", function() {
	    deferred.resolve(body);
	});

    });
    return deferred.promise
}
exports.downloadBody = downloadBody;

function downloadSubReddit(subreddit) {
    var url = "http://www.reddit.com/r/" + subreddit + ".json";
    return downloadBody(url).then(parseSubReddit);
}
exports.downloadSubReddit =  downloadSubReddit

function downloadFirst(permalink) {
    var url = "http://www.reddit.com" + permalink + ".json";
    return downloadBody(url).then(parseFirst);
}
exports.downloadFirst = downloadFirst;

/** 
    Parsers
**/
function parseSubReddit(body) {
    var permalinksM = lift(maybe, permalinks);

    return permalinksM(
	children(JSON.parse(body))
    );
}
exports.parseSubReddit = parseSubReddit;

function parseFirst(body) {
    var listing = JSON.parse(body);
    var firstM = lift(maybe, first);
    return firstM(comments(listing)).bind(makeComment);
}
exports.parseFirst = parseFirst;

/**
   Accessors
**/
function makeComment(entity) {
    return maybe(entity.data).bind(
	function(data) {
	    var body = data.body;
	    var created = data.created;
	    return body && created 
		? maybe({body: body, created: created}) 
	        : maybe(null);
	})
}

function first(list) {
    return list[0];
}

function comments(listing) {
    return maybe(listing[1]).bind(children)
}

function children(entity) {
    return getNested(entity, "data", "children");
}

function permalinks(children) {
    return children.reduce(
	function(arr, entity) {
	    permalink(entity).bind(function(x) {
		arr.push(x);
	    });
	    return arr;
	},
	[]);
}

function permalink(entity) {
    return getNested(entity, "data", "permalink");
}

/**
   Utilities
**/
function lift(monad, f) {
    return function(m) {
	return m.bind(function(x) { return monad(f(x))});
    }
}
exports.lift = lift;


function maybe(x) {
   return {
       bind: function(f) { 
	   if(x == null) {
	       return maybe(null);
	   } else {
	       return f(x);
	   }
       },
       getDefault: function(y) {
	   return x == null ? y : x
       }
   }
}
exports.maybe = maybe;

function lookup(key, obj) {
    return function(obj) {
	return maybe(obj[key]);
    };
};

function getNested(obj) {
    var keys = [].slice.call(arguments, 1);
    return [].reduce.call(keys,
	function(m, key) {
	    return m.bind(lookup(key));
	},
	maybe(obj)
    );
}

File: haskell/firsts.hs
{-# LANGUAGE DeriveDataTypeable #-}

module Main where

import Network.HTTP
import Control.Monad
import Text.JSON
import Text.JSON.Types
import Text.JSON.Generic
import Control.Concurrent.Async
import System.Environment
import Data.List (sort)
import Data.Time
import Debug.Trace (trace, traceShow)

data Comment = Comment {
  body :: String,
  created :: Int
} deriving (Show, Eq)

instance Ord Comment where
  compare x y = descending $ compare (created x) (created y)

descending :: Ordering -> Ordering
descending LT = GT
descending GT = LT
descending EQ = EQ

{-==============================================================================
JSON processors
==============================================================================-}
-- Converts the subreddit data into a list of permalinks
parseSubReddit :: String -> Result [String]
parseSubReddit = decode >=> children >=> return . permalinks

-- Converts the post data into the first comment
parseFirst :: String -> Result [Comment]
parseFirst = decode >=> comments >=> return . firstComment

{-==============================================================================
#  Entity accessors
==============================================================================-}
children :: JSObject JSValue -> Result [JSObject JSValue]
children = valFromObj "data" >=> valFromObj "children"

permalinks :: [JSObject JSValue] -> [String]
permalinks = concatMap (resultToList . permalink)

permalink :: JSObject JSValue -> Result String
permalink = valFromObj "data" >=> valFromObj "permalink"

-- Converts a comment into one or zero comments (zero if invalid)
comment :: JSObject JSValue -> [Comment]
comment obj = resultToList $ do
  dat <- valFromObj "data" obj
  body <- valFromObj "body" dat
  created <- valFromObj "created" dat
  return Comment{body=body, created=created}

comments :: [JSObject JSValue] -> Result [Comment]
comments (_:commentListing:_) =
  valFromObj "data" commentListing >>= valFromObj "children" >>= return . concatMap comment

firstComment :: [Comment] -> [Comment]
firstComment = take 1 

{-==============================================================================
# IO 
==============================================================================-}
data Mode = Async | Sync

downloadAll mode subreddits = do
  links <- concat `liftM` mapper downloadAndParseSubReddit subreddits
  comments <- concat `liftM` mapper downloadAndParseFirstComment links
  return $ sort comments
  where
    mapper = case mode of
      Async -> mapConcurrently
      Sync  -> mapM
    -- Note: the concat is here because paresSubReddit return a Result [String]
    --       resultToList turns that into [[String]]
    downloadAndParseSubReddit    = liftM (concat . resultToList . parseSubReddit) . downloadSubReddit
    downloadAndParseFirstComment = liftM (concat . resultToList . parseFirst) . downloadPost

tc f = do
  start <- getCurrentTime
  ret <- f
  end <- getCurrentTime
  return (diffUTCTime end start, ret)

main = print . fst <=< tc $ do
  subreddits <- getArgs
  mode <- getMode `liftM` getProgName 
  permalinks <- downloadAll mode subreddits
  return ()
  where
    getMode "firsts-async" = Async
    getMode "firsts-sync"  = Sync
{-==============================================================================
# Utilities
==============================================================================-}
resultToList (Ok v)    = [v]
resultToList (Error _) = [ ]

redditURL = "http://www.reddit.com"
subredditURL subreddit = redditURL ++ "/r/" ++ subreddit ++ ".json"
postURL permalink = redditURL ++ permalink ++ ".json"


downloadBody = getResponseBody <=< simpleHTTP . getRequest
downloadSubReddit = downloadBody . subredditURL
downloadPost :: String -> IO String
downloadPost = downloadBody . postURL

traceArg msg x =
  trace msg' x
  where
    msg' = msg ++ ": " ++ show x


